File: src-angular/task-manager/src/app/data/in-memory-task-repository.ts
import { Injectable } from '@angular/core';
import { Task } from '../domain/task';
import { TaskId } from '../domain/task-id';
import { TaskRepository } from '../domain/task-repository';
import { TaskStatus } from '../domain/task-status';

/**
 * In-memory implementation of TaskRepository, used for the workshop labs.
 * Mirrors the "in-memory repository" pattern used in the .NET/Spring Boot/etc. tracks'
 * Infrastructure layer.
 */
@Injectable({ providedIn: 'root' })
export class InMemoryTaskRepository implements TaskRepository {
  private readonly tasks = new Map<string, Task>();

  // TODO: Participants will implement these methods during the workshop

  async findById(taskId: TaskId): Promise<Task | undefined> {
    throw new Error('Not implemented: findById');
  }

  async getActiveTasks(): Promise<Task[]> {
    throw new Error('Not implemented: getActiveTasks');
  }

  async addTask(task: Task): Promise<void> {
    throw new Error('Not implemented: addTask');
  }

  async saveChanges(task: Task): Promise<void> {
    throw new Error('Not implemented: saveChanges');
  }
}

File: src-python/task_manager_infrastructure/in_memory_task_repository.py
"""In-memory repository implementation for workshop purposes."""

from __future__ import annotations

from task_manager_domain import Task, TaskId, TaskStatus


class InMemoryTaskRepository:
    """Store tasks in memory using a plain dictionary."""

    def __init__(self) -> None:
        self._tasks: dict[TaskId, Task] = {}

    async def find_by_id(self, task_id: TaskId) -> Task | None:
        return self._tasks.get(task_id)

    async def get_active_tasks(self) -> list[Task]:
        return [
            task
            for task in self._tasks.values()
            if task.status not in {TaskStatus.DONE, TaskStatus.CANCELLED}
        ]

    async def add_task(self, task: Task) -> None:
        self._tasks[task.id] = task

    async def save_changes(self, task: Task) -> None:
        self._tasks[task.id] = task
